### ISSUE TITLE
Custom class need to be declared inside layer directive 

### ISSUE THREAD

**Issue by danielweil** [tailwindlabs/tailwindcss#5989]
I am getting an error randomly on my NextJS application where it gives a message telling me I have to declare a custom class inside a @layer directive, even if I am already doing so. I say random, bacause if I hard reset to HEAD, it sometimes goes away. My problem is that now it started giving the error inside Azure Pipelines.

![image](https://user-images.githubusercontent.com/44274213/140448312-2aac5423-1d74-461e-89db-b0e753c5cccb.png)

![image](https://user-images.githubusercontent.com/44274213/140448318-d2f50106-3b2a-41db-957b-6017ae2ab390.png)

**Comment by DennisKraaijeveld** [comment]
https://github.com/gatsbyjs/gatsby/issues/33688 Bug is in Gatsby as well. Tailwind is breaking my themes more and more. I loved it at first, but this is getting really counterproductive

**Comment by adamwathan** [comment]
Hey! Please provide a minimal reproduction or it is impossible to help.

**Comment by DennisKraaijeveld** [comment]
@adamwathan there is actually a reproduction link

https://github.com/sklawren/gatsby-tailwind-issue

Only difference is the framework (Gatsby vs Next.js), but the issue is the same since both are using .module.css files and the build will just crash.

**Comment by danielweil** [comment]
> Hey! Please provide a minimal reproduction or it is impossible to help.

I don't know if I can reproduce because I don't know what is causing it. It is simply random. I just recreated the Azure pipeline and the error disappeared.

When it happens in my local app, I have to hard reset to it magically start working again.

**Comment by adamwathan** [comment]
@danielweil Looking at your screenshot I think I have a guess at the problem — you're using CSS modules right (which isn't mentioned in your issue but would have been helpful to know! 😅) and CSS modules are always processed independently by build tools.

What this means is that Tailwind is running from scratch in isolation on each CSS module, and on your global.css file. This means each run sort of has no implicit knowledge of other runs, so classes defined in global.css are not "visible" to your CSS modules, again because Tailwind is running in total isolation for each file, as if each file were on a separate computer in different countries.

This is sort of annoying so we actually _try_ to share knowledge between these builds as best we can by using some heuristics to guess if two separate builds should know about each other and share a context. What I think is happening is that PostCSS is running against your multiple CSS files but _not in a guaranteed order_. When `global.css` is processed first, everything is fine. But if another file is processed first, the classes defined in `global.css` haven't been picked up yet, and the build fails.

I'm not sure what we can do to handle this better honestly — probably the best thing we can do is error _every_ time since we can't do anything to guarantee it works. Unfortunately this is pretty out of our hands, as we can't control how tools like webpack process CSS. 

**The good news is there is a solution** — you just need to define any classes you want to `@apply` in your CSS modules in your `tailwind.config.js` file as plugins instead of defining them in your `global.css` file. Then those classes are guaranteed to be visible to any CSS file that is using the same Tailwind config:

```js
module.exports = {
  // ...
  plugins: [
    function ({ addUtilities }) {
      addUtilities({
        '.responsive-text': {
          '@apply text-xs lg:text-tiny xl:text-base 2xl:text-lg': {},
        },
        '.responsive-text-lg': {
          '@apply text-xs lg:text-base xl:text-lg 2xl:text-xl': {},
        }
        // ...
      })
    }
  ]
}
```

Admittedly it's a little annoying compared to just authoring it in your CSS file, but putting this code in your config file is the only way to guarantee it's visible to every CSS file being compiled with Tailwind that uses this file.

Again, what is happening is that your build tool is effectively running all of these processes in parallel:

```
tailwindcss -i global.css
tailwindcss -i module-one.module.css
tailwindcss -i module-two.module.css
tailwindcss -i module-three.module.css
```

When you realize this it is a bit less surprising that code in `module-one.module.css` can't "see" code in `global.css`. But if they share a config file and the code is in the config file, you are all good.

Let me know if that solves the issue for you, and if so will close since unfortunately there's nothing we can do code-wise to improve this.

**Comment by stevensacks** [comment]
@adamwathan Thank you for the detailed explanation.

I’m still left wondering why this issue does not occur in Gatsby 2 or 3, but does in Gatsby 4.

**Comment by danielweil** [comment]
> @danielweil Looking at your screenshot I think I have a guess at the problem — you're using CSS modules right (which isn't mentioned in your issue but would have been helpful to know! 😅) and CSS modules are always processed independently by build tools.
> 
> What this means is that Tailwind is running from scratch in isolation on each CSS module, and on your global.css file. This means each run sort of has no implicit knowledge of other runs, so classes defined in global.css are not "visible" to your CSS modules, again because Tailwind is running in total isolation for each file, as if each file were on a separate computer in different countries.
> 
> Now for performance reasons we actually _try_ to share knowledge between these builds as best we can by using some heuristics to guess if two separate builds should share what is internally known as a "context". What I think is happening is that PostCSS is running against your multiple CSS files but _not in a guaranteed order_. When `global.css` is processed first, the classes in it are picked up and stored in the context and everything is fine. But if another file is processed first, the classes defined in `global.css` haven't been picked up yet, and the build fails.
> 
> I'm not sure what we can do to handle this better honestly — probably the best thing we can do is error _every_ time since we can't do anything to guarantee it works. Unfortunately this is pretty out of our hands, as we can't control how tools like webpack process CSS.
> 
> **The good news is there is a solution** — you just need to define any classes you want to `@apply` in your CSS modules in your `tailwind.config.js` file as plugins instead of defining them in your `global.css` file. Then those classes are guaranteed to be visible to any CSS file that is using the same Tailwind config:
> 
> ```js
> module.exports = {
>   // ...
>   plugins: [
>     function ({ addUtilities }) {
>       addUtilities({
>         '.responsive-text': {
>           '@apply text-xs lg:text-tiny xl:text-base 2xl:text-lg': {},
>         },
>         '.responsive-text-lg': {
>           '@apply text-xs lg:text-base xl:text-lg 2xl:text-xl': {},
>         }
>         // ...
>       })
>     }
>   ]
> }
> ```
> 
> Admittedly it's a little annoying compared to just authoring it in your CSS file, but putting this code in your config file is the only way to guarantee it's visible to every CSS file being compiled with Tailwind that uses this file.
> 
> Again, what is happening is that your build tool is effectively running all of these processes in parallel:
> 
> ```
> tailwindcss -i global.css
> tailwindcss -i module-one.module.css
> tailwindcss -i module-two.module.css
> tailwindcss -i module-three.module.css
> ```
> 
> When you realize this it is a bit less surprising that code in `module-one.module.css` can't "see" code in `global.css`. But if they share a config file and the code is in the config file, you are all good.
> 
> Let me know if that solves the issue for you, and if so will close since unfortunately there's nothing we can do code-wise to improve this.

Perfect. Will try and let you know if it works out!

**Comment by eddyharold** [comment]
Hello
I have the same issue but in my case I'm using React, not Nextjs
Here is the error i receive when i compile my projet
![Windows PowerShell 11_11_2021 11_18_14 AM (2)](https://user-images.githubusercontent.com/36563793/141281318-d338e16c-bd72-4955-b19d-07cb7a761278.png)

Please, really need some help
Thanks

**Comment by RobinMalfait** [comment]
@eddyharold according to the error message that class doesn't exist. Which probably means that primary-300 doesn't exist. Please open a new issue with a **minimal** reproduction repo so that we can take a look.

**Comment by DennisKraaijeveld** [comment]
@RobinMalfait Thanks for your response. But the class isn't there, exactly for the same reason as our issues. Because it will not pickup the custom classes added with the @layer directive. First this did work. Right now we have the same issue on Gatsby, React and Next, so what did actually change?

**Comment by eddyharold** [comment]
> @eddyharold according to the error message that class doesn't exist. Which probably means that primary-300 doesn't exist. Please open a new issue with a **minimal** reproduction repo so that we can take a look.

Thanks for your reply @RobinMalfait , but I finally solved my problem.
My app was implementing the dark mode. On my sccs, I was customizing some classes to synchronize with the theme mode.
In my case, the error was appearring when I disabled dark mode in my tailwind configurations.
So, I re-enabled the dark mode and the error disappeared.

**Comment by adamwathan** [comment]
Going to close this as unfortunately absolutely nothing we can do about it — the solution would be for Gatsby to process all CSS modules without Tailwind, then concatenate them into one stylesheet, and pass that stylesheet to Tailwind so Tailwind is only run once with a shared context. There's no way for us to do anything about that, it absolutely has to be done at the build tool level and I don't except any build tools to make a drastic change like that. I'm not sure why it would have worked before and not now — my only thought is that something inside Gatsby was always processing `global.css` first and now they are doing more in parallel or something.

This is documented here at the moment:

https://tailwindcss.com/docs/functions-and-directives#using-apply-with-per-component-css

Will make a note to update the documentation to mention CSS modules here as well since the problem is the same.

**Comment by joggienl** [comment]
Hi @adamwathan, I got this same behaviour today after upgrading from tailwindcss v2 to v3. It is happening in a NextJS project, where we use css modules and postcss syntax. So similar to this issues topic.

I've isolated the issue in an example, perhaps that is interesting as wel: https://stackblitz.com/edit/nextjs-p3g1v7 (note, stackblitz works on chrome only currently if I am correct).

As far as I know this worked before with tailwindcss v2, but I cannot confirm that it might be by accident due to ordering of rendering as you noted. Perhaps it is worth a note in the upgrade docs?

about the link https://tailwindcss.com/docs/functions-and-directives#using-apply-with-per-component-css, there is an error in the example of the tailwind.config.js

```js
const plugin = require('tailwindcss/plugin')

module.exports = {
  // ...
  plugins: [
    plugin(function ({ addComponents, theme }) {
      addComponents({
        '.card': {
          backgroundColor: theme('colors.white');
          borderRadius: theme('borderRadius.lg');
          padding: theme('spacing.6');
          boxShadow: theme('boxShadow.xl');
        }
      })
    })
  ]
}
```
the semicolons in the `'.card'` object should be commas 😅

**Comment by eybarta** [comment]
@adamwathan 

Hi,
What if I wanted to use custom classes that can't use the @apply on all their props such as this:
`
.rp-scroll {
    overflow: auto;
    &::-webkit-scrollbar-track {
        background-color: transparent;
    }
    &::-webkit-scrollbar-thumb {
        @apply rounded bg-gray-600;
    }
    &::-webkit-scrollbar {
        width: 6px;
        height: 4px;
        left: 10px;
    }
    &.thin {
        &::-webkit-scrollbar {
            width: 2px;
        }
    }
}
`

How would I add that into the tailwind.config plugins ? Thx!
